proposals/texture-buffers.md

```diff
@@ -0,0 +1,405 @@
+
+```
+R8G8B8A8_UNORM
+R8G8B8A8_UINT
+R8G8B8A8_SINT
+R16G16B16A16_UINT
+R16G16B16A16_SINT
+R16G16B16A16_SFLOAT
+R32_UINT
+R32_SINT
+R32_SFLOAT
+R32G32_UINT
+R32G32_SINT
+R32G32_SFLOAT
+R32G32B32A32_UINT
+R32G32B32A32_SINT
+R32G32B32A32_SFLOAT
+```
+
+For the other formats, [gpuinfo.org](https://vulkan.gpuinfo.org/listbufferformats.php) has information on how widespread support is.
+For 1- and 2-channel `R{8,16}_{SINT,UINT,SFLOAT}`, support is currently around 80% for storage texel buffers.
+
+**TODO:** We should do this query against WebGPU's baseline requirements, as the percentage for devices we actually support may be higher.
+
+Vulkan has a `maxTexelBufferElements` limit for the maximum size of a texel buffer.
+[gpuinfo.org shows that](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxTexelBufferElements&platform=all) more than 85% of devices support 128MB texel buffers.
+
+
+### Metal
+
+Metal has a `texture_buffer` type that provides similar functionality, which was introduced in Metal 2.1.
+The [Metal Feature Set Tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) show the supported formats for each access mode.
+Unnormalized integer and floating point formats are supported for all access modes, as is `RGBA8Unorm`.
+Using a `read-write` access mode also requires support for the Tier 2 `MTLReadWriteTextureTier`.
+A `mem_texture` fence would be needed to make texture buffer writes visible within an invocation.
+
+To get coverage on older Metal versions, it would be possible to polyfill by using a regular device buffer and doing the format conversions inside the shader.
+This requires that the storage format is specified inside the shader.
+
+The maximum texture buffer size is 64MB for the Apple2 GPU family, and 256MB for Apple3 and above.
+
+
+### D3D12
+
+In D3D12, a texture buffer can map to an Unordered Access View (UAV) for a buffer with a `DXGI_FORMAT`, and that UAV can be accessed in the shader with 32-bit result types.
+See [Typed unordered access view (UAV) loads](https://docs.microsoft.com/en-us/windows/win32/direct3d12/typed-unordered-access-view-loads).
+The `RWByteAddressBuffer` should be prefixed with `globallycoherent`.
+
+Format support for typed UAV loads and stores in D3D12 can be checked [here](https://docs.microsoft.com/en-us/windows/win32/direct3ddxgi/hardware-support-for-direct3d-12-0-formats).
+The set of required formats includes:
+
+```
+R8G8B8A8_UNORM
+R8G8B8A8_UINT
+R8G8B8A8_SINT
+R16G16B16A16_UINT
+R16G16B16A16_SINT
+R16G16B16A16_SFLOAT
+R8_UINT
+R8_SINT
+R32_UINT
+R32_SINT
+R32_SFLOAT
+R32G32B32A32_UINT
+R32G32B32A32_SINT
+R32G32B32A32_SFLOAT
+```
+
+
+# Open Questions
+
+1. Should this be an extension, or a core feature?
+   - To make it core, implementations would need to polyfill for Metal <2.1. We would also need to drop the formats that are not required everywhere (e.g. `R8_UINT`), or make them optional.
```

Comment:
We already have `bgra8unorm-storage` and `rg11b10ufloat-renderable` so assuming the polyfill in older Metal isn't too bad a core feature with optional format support would be a good choice I think.

Reply:
I agree it would be nice to have a core feature.
I don't have a clear picture of which features would have to drop out.

A quick look at the table for Apple GPUs, for example, shows `RG32Float` supports Read, Write, but not read-write.

checking equivalents to the vulkan ones, "all" means read, write, and read-write are supported:

* R8G8B8A8_UNORM, "all"
* R8G8B8A8_UINT, "all"
* R8G8B8A8_SINT, "all"
* R16G16B16A16_UINT, "all"
* R16G16B16A16_SINT, "all"
* R16G16B16A16_SFLOAT, "all"
* R8_UINT, "all"
* R8_SINT, "all"
* R32_UINT, "all"
* R32_SINT, "all"
* R32_SFLOAT, "all"
* R32G32B32A32_UINT.  "all"
* R32G32B32A32_SINT.  "all"
* R32G32B32A32_SFLOAT, "all"

So that's actually no restriction for Apple GPUs.

The table showing maximum texture buffer width only has columns for Apple2  + GPUs, and also "Mac2" which ... I can't decipher.

Reply:
Made it a core feature as decided at the F2F. The optional formats are left for a future texture format tier extension.